[hop][print]Add args print support to hop print

fxdawnn opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #170880
* #167680

## Problem
The previous hop print only support kwargs print instead of the general format print that consider both args and kwargs.

## Solutions
This Higher Order Operator (HOP) provides a functional version of print reusing the previous implemented kwargs infrastructure and extended from only supporting 1 to both 1, 2, and 3.
    It supports the calling conventions of print(format_str.format(*args, **kwargs)):

    1. Format string with keyword arguments (named placeholders):
       torch._higher_order_ops.print("moo {x} {y}", x=1, y=2)
       Output: "moo 1 2"

    2. Format string with positional arguments (positional placeholders):
       torch._higher_order_ops.print("moo {} {}", 1, 2)
       Output: "moo 1 2"

    3. Mixed positional and keyword arguments:
       torch._higher_order_ops.print("moo {} {y}", 1, y=2)
       Output: "moo 1 2"

    This HOP enables printing without causing graph break.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @mlazos

fxdawnn committed:
[hop][print]Add args print support to hop print

[ghstack-poisoned]

angelayi commented on torch/_higher_order_ops/print.py:
```diff
@@ -47,48 +65,57 @@ def gen_schema(self, format_str: str, **kwargs: object) -> torch.FunctionSchema:
         lambda a: map_types[type(a)](a),
         kwargs,
         lambda a: isinstance(a, tuple(map_types.keys())),
     )
```
```suggestion
    new_args, new_kwargs = pytree.tree_map_only(
        tuple(map_types.keys()),
        lambda a: map_types[type(a)](a),
        (args, kwargs),
        lambda a: isinstance(a, tuple(map_types.keys())),
    )
```

fxdawnn committed:
Update on "[hop][print]Add args print support to hop print"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

fxdawnn commented on torch/_higher_order_ops/print.py:
```diff
@@ -9,34 +9,52 @@
 
     This Higher Order Operator (HOP) provides a functional version of print for use in PyTorch graphs.
-    It enables format printing with named arguments, e.g., torch._higher_order_ops.print("moo {x} {y}", x=1, y=2).
+    It supports two calling conventions:
```
updated. Thanks

fxdawnn committed:
Update on "[hop][print]Add args print support to hop print"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

fxdawnn committed:
Update on "[hop][print]Add args print support to hop print"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

fxdawnn committed:
Update on "[hop][print]Add args print support to hop print"


## Problem
The previous hop print only support kwargs print instead of the general format print that consider both args and kwargs.

## Solutions
This Higher Order Operator (HOP) provides a functional version of print reusing the previous implemented kwargs infrastructure and extended from only supporting 1 to both 1, 2, and 3.
    It supports the calling conventions of print(format_str.format(*args, **kwargs)):

    1. Format string with keyword arguments (named placeholders):
       torch._higher_order_ops.print("moo {x} {y}", x=1, y=2)
       Output: "moo 1 2"

    2. Format string with positional arguments (positional placeholders):
       torch._higher_order_ops.print("moo {} {}", 1, 2)
       Output: "moo 1 2"

    3. Mixed positional and keyword arguments:
       torch._higher_order_ops.print("moo {} {y}", 1, y=2)
       Output: "moo 1 2"

    This HOP enables printing without causing graph break.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

fxdawnn committed:
Update on "[hop][print]Add args print support to hop print"


## Problem
The previous hop print only support kwargs print instead of the general format print that consider both args and kwargs.

## Solutions
This Higher Order Operator (HOP) provides a functional version of print reusing the previous implemented kwargs infrastructure and extended from only supporting 1 to both 1, 2, and 3.
    It supports the calling conventions of print(format_str.format(*args, **kwargs)):

    1. Format string with keyword arguments (named placeholders):
       torch._higher_order_ops.print("moo {x} {y}", x=1, y=2)
       Output: "moo 1 2"

    2. Format string with positional arguments (positional placeholders):
       torch._higher_order_ops.print("moo {} {}", 1, 2)
       Output: "moo 1 2"

    3. Mixed positional and keyword arguments:
       torch._higher_order_ops.print("moo {} {y}", 1, y=2)
       Output: "moo 1 2"

    This HOP enables printing without causing graph break.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

fxdawnn committed:
Update on "[hop][print]Add args print support to hop print"


## Problem
The previous hop print only support kwargs print instead of the general format print that consider both args and kwargs.

## Solutions
This Higher Order Operator (HOP) provides a functional version of print reusing the previous implemented kwargs infrastructure and extended from only supporting 1 to both 1, 2, and 3.
    It supports the calling conventions of print(format_str.format(*args, **kwargs)):

    1. Format string with keyword arguments (named placeholders):
       torch._higher_order_ops.print("moo {x} {y}", x=1, y=2)
       Output: "moo 1 2"

    2. Format string with positional arguments (positional placeholders):
       torch._higher_order_ops.print("moo {} {}", 1, 2)
       Output: "moo 1 2"

    3. Mixed positional and keyword arguments:
       torch._higher_order_ops.print("moo {} {y}", 1, y=2)
       Output: "moo 1 2"

    This HOP enables printing without causing graph break.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

fxdawnn commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1580,11 +1591,20 @@ def _generate_print_fallback(self, node: ir.FallbackKernel) -> None:
+            elif args_str:
+                self.writeline(f"print({repr(format_str)}.format({args_str}))")
+            else:
+                self.writeline(f"print({repr(format_str)}.format({kwargs_str}))")
```
updated in the new ``custom_codegen``

fxdawnn committed:
Update on "[hop][print]Add args print support to hop print"


## Problem
The previous hop print only support kwargs print instead of the general format print that consider both args and kwargs.

## Solutions
This Higher Order Operator (HOP) provides a functional version of print reusing the previous implemented kwargs infrastructure and extended from only supporting 1 to both 1, 2, and 3.
    It supports the calling conventions of print(format_str.format(*args, **kwargs)):

    1. Format string with keyword arguments (named placeholders):
       torch._higher_order_ops.print("moo {x} {y}", x=1, y=2)
       Output: "moo 1 2"

    2. Format string with positional arguments (positional placeholders):
       torch._higher_order_ops.print("moo {} {}", 1, 2)
       Output: "moo 1 2"

    3. Mixed positional and keyword arguments:
       torch._higher_order_ops.print("moo {} {y}", 1, y=2)
       Output: "moo 1 2"

    This HOP enables printing without causing graph break.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

fxdawnn committed:
Update torch/_inductor/codegen/custom_extern_kernel_codegen.py

Co-authored-by: Angela Yi <yiangela7@gmail.com>

fxdawnn committed:
Update on "[hop][print]Add args print support to hop print"


## Problem
The previous hop print only support kwargs print instead of the general format print that consider both args and kwargs.

## Solutions
This Higher Order Operator (HOP) provides a functional version of print reusing the previous implemented kwargs infrastructure and extended from only supporting 1 to both 1, 2, and 3.
    It supports the calling conventions of print(format_str.format(*args, **kwargs)):

    1. Format string with keyword arguments (named placeholders):
       torch._higher_order_ops.print("moo {x} {y}", x=1, y=2)
       Output: "moo 1 2"

    2. Format string with positional arguments (positional placeholders):
       torch._higher_order_ops.print("moo {} {}", 1, 2)
       Output: "moo 1 2"

    3. Mixed positional and keyword arguments:
       torch._higher_order_ops.print("moo {} {y}", 1, y=2)
       Output: "moo 1 2"

    This HOP enables printing without causing graph break.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

fxdawnn committed:
Update on "[hop][print]Add args print support to hop print"


## Problem
The previous hop print only support kwargs print instead of the general format print that consider both args and kwargs.

## Solutions
This Higher Order Operator (HOP) provides a functional version of print reusing the previous implemented kwargs infrastructure and extended from only supporting 1 to both 1, 2, and 3.
    It supports the calling conventions of print(format_str.format(*args, **kwargs)):

    1. Format string with keyword arguments (named placeholders):
       torch._higher_order_ops.print("moo {x} {y}", x=1, y=2)
       Output: "moo 1 2"

    2. Format string with positional arguments (positional placeholders):
       torch._higher_order_ops.print("moo {} {}", 1, 2)
       Output: "moo 1 2"

    3. Mixed positional and keyword arguments:
       torch._higher_order_ops.print("moo {} {y}", 1, y=2)
       Output: "moo 1 2"

    This HOP enables printing without causing graph break.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

fxdawnn committed:
Update on "[hop][print]Add args print support to hop print"


## Problem
The previous hop print only support kwargs print instead of the general format print that consider both args and kwargs.

## Solutions
This Higher Order Operator (HOP) provides a functional version of print reusing the previous implemented kwargs infrastructure and extended from only supporting 1 to both 1, 2, and 3.
    It supports the calling conventions of print(format_str.format(*args, **kwargs)):

    1. Format string with keyword arguments (named placeholders):
       torch._higher_order_ops.print("moo {x} {y}", x=1, y=2)
       Output: "moo 1 2"

    2. Format string with positional arguments (positional placeholders):
       torch._higher_order_ops.print("moo {} {}", 1, 2)
       Output: "moo 1 2"

    3. Mixed positional and keyword arguments:
       torch._higher_order_ops.print("moo {} {y}", 1, y=2)
       Output: "moo 1 2"

    This HOP enables printing without causing graph break.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

fxdawnn commented on test/higher_order_ops/test_print.py:
```diff
@@ -32,40 +32,88 @@ def f(x):
+            x = x + x
+            torch._higher_order_ops.print("moo {} {y}", 1, y=2)
+            x = x * x
+            return x
```
merged the tests

fxdawnn committed:
Update on "[hop][print]Add args print support to hop print"


## Problem
The previous hop print only support kwargs print instead of the general format print that consider both args and kwargs.

## Solutions
This Higher Order Operator (HOP) provides a functional version of print reusing the previous implemented kwargs infrastructure and extended from only supporting 1 to both 1, 2, and 3.
    It supports the calling conventions of print(format_str.format(*args, **kwargs)):

    1. Format string with keyword arguments (named placeholders):
       torch._higher_order_ops.print("moo {x} {y}", x=1, y=2)
       Output: "moo 1 2"

    2. Format string with positional arguments (positional placeholders):
       torch._higher_order_ops.print("moo {} {}", 1, 2)
       Output: "moo 1 2"

    3. Mixed positional and keyword arguments:
       torch._higher_order_ops.print("moo {} {y}", 1, y=2)
       Output: "moo 1 2"

    This HOP enables printing without causing graph break.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

fxdawnn committed:
Update on "[hop][print]Add args print support to hop print"


## Problem
The previous hop print only support kwargs print instead of the general format print that consider both args and kwargs.

## Solutions
This Higher Order Operator (HOP) provides a functional version of print reusing the previous implemented kwargs infrastructure and extended from only supporting 1 to both 1, 2, and 3.
    It supports the calling conventions of print(format_str.format(*args, **kwargs)):

    1. Format string with keyword arguments (named placeholders):
       torch._higher_order_ops.print("moo {x} {y}", x=1, y=2)
       Output: "moo 1 2"

    2. Format string with positional arguments (positional placeholders):
       torch._higher_order_ops.print("moo {} {}", 1, 2)
       Output: "moo 1 2"

    3. Mixed positional and keyword arguments:
       torch._higher_order_ops.print("moo {} {y}", 1, y=2)
       Output: "moo 1 2"

    This HOP enables printing without causing graph break.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

fxdawnn committed:
Update on "[hop][print]Add args print support to hop print"


## Problem
The previous hop print only support kwargs print instead of the general format print that consider both args and kwargs.

## Solutions
This Higher Order Operator (HOP) provides a functional version of print reusing the previous implemented kwargs infrastructure and extended from only supporting 1 to both 1, 2, and 3.
    It supports the calling conventions of print(format_str.format(*args, **kwargs)):

    1. Format string with keyword arguments (named placeholders):
       torch._higher_order_ops.print("moo {x} {y}", x=1, y=2)
       Output: "moo 1 2"

    2. Format string with positional arguments (positional placeholders):
       torch._higher_order_ops.print("moo {} {}", 1, 2)
       Output: "moo 1 2"

    3. Mixed positional and keyword arguments:
       torch._higher_order_ops.print("moo {} {y}", 1, y=2)
       Output: "moo 1 2"

    This HOP enables printing without causing graph break.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

fxdawnn committed:
Update on "[hop][print]Add args print support to hop print"


## Problem
The previous hop print only support kwargs print instead of the general format print that consider both args and kwargs.

## Solutions
This Higher Order Operator (HOP) provides a functional version of print reusing the previous implemented kwargs infrastructure and extended from only supporting 1 to both 1, 2, and 3.
    It supports the calling conventions of print(format_str.format(*args, **kwargs)):

    1. Format string with keyword arguments (named placeholders):
       torch._higher_order_ops.print("moo {x} {y}", x=1, y=2)
       Output: "moo 1 2"

    2. Format string with positional arguments (positional placeholders):
       torch._higher_order_ops.print("moo {} {}", 1, 2)
       Output: "moo 1 2"

    3. Mixed positional and keyword arguments:
       torch._higher_order_ops.print("moo {} {y}", 1, y=2)
       Output: "moo 1 2"

    This HOP enables printing without causing graph break.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

fxdawnn commented:
@pytorchbot merge
